### PR TITLE
BUG: groupby().rolling(freq) with monotonic dates within groups

### LINKED ISSUE #46061: BUG: groupby().rolling("freq") validation raises even if dates are sorted within each group
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
print(pd.__version__)
test_df = pd.DataFrame({
    'Name':['L', 'L', 'L',
                    'M', 'M', 'M', 'M', 'M', 'M'],
    'ID': ['L1', 'L2', 'L3',
              'M1', 'M2', 'M3', 'M4', 'M5', 'M6'],
    'Date': [
        '2005-01-01', '2006-01-01', '2013-01-01', 
        '2005-01-01', '2006-01-01', '2006-01-01', '2017-01-01', '2006-02-01', '2006-03-01'],
    'indicator': [0, 1, 0,
                          0, 1, 0, 0, 1, 0],
    'rolling_sum':[0, 0, 1,
                              0, 0, 0, 0, 1, 2]})
test_df['Date'] = pd.to_datetime(test_df['Date'], errors='coerce')
test_df.groupby('Name').rolling('731D', on='Date')['indicator'].sum().reset_index()
```

### Issue Description

Grouping by a column and then rolling over the grouped dataframe results in `[ValueError: Date must be monotonic]()`. Ultimately, the operation should be able to produce the `rolling_sum` column in the example.

### Expected Behavior

This is not an issue in pandas `1.3.5` but is an issue in `1.4` and higher. I expect this function to create a grouped dataframe and to then perform the rolling summation on the provided column using the date window provided. 

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.8.12.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.1.0
Version          : Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:24 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T8101
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.4.1
numpy            : 1.22.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 22.0.3
setuptools       : 60.6.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 8.0.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 6.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #46061
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/v1.4.2.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by rhshadrach — COMMENTED** [review]
lgtm, does the iteration in _validate_datetimelike_monotonic have a big performance impact?

**Comment by mroeschke** [comment]
There will be a slight performance impact during the validation (that can be a 1 time hit) that scales with the number of groups. Here's validating 3 vs 100 groups

```
In [1]:
   ...:         dates = date_range(start="2016-01-01 09:30:00", periods=20, freq="s")
   ...:         df = DataFrame(
   ...:             {
   ...:                 "A": [1] * 20 + [2] * 12 + [3] * 8,
   ...:                 "B": np.concatenate((dates, dates)),
   ...:                 "C": np.arange(40),
   ...:             }
   ...:         )

In [2]: %timeit df.groupby("A").rolling("4s", on="B")
764 µs ± 38.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]:
   ...:         dates = date_range(start="2016-01-01 09:30:00", periods=50, freq="s")
   ...:         df = DataFrame(
   ...:             {
   ...:                 "A": np.arange(100),
   ...:                 "B": np.concatenate((dates, dates)),
   ...:                 "C": np.arange(100),
   ...:             }
   ...:         )

In [7]: %timeit df.groupby("A").rolling("4s", on="B")
2.26 ms ± 30.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by rhshadrach** [comment]
@mroeschke - It looks like the call to `groupby(...).rolling(...)` went from O(1) to O(n), is that right? I would describe that as being something significant. Perhaps it can't be avoided, but I'd like to have another look. In any case, I don't think we should be merging PRs (especially our own) when there are outstanding questions.

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -2631,12 +2626,17 @@ def _get_window_indexer(self) -> GroupbyIndexer:
...
-            not (self._on.is_monotonic_increasing or self._on.is_monotonic_decreasing)
-            or self._on.hasnans
-        ):
-            self._raise_monotonic_error()
+        # GH 46061
+        on = "index" if self.on is None else self.on
+        if self._on.hasnans:
+            raise ValueError(f"{on} must not have any NaT values.")
+        for group_indices in self._grouper.indices.values():
+            group_on = self._on.take(group_indices)
```
Comment (jreback) [inline pandas/core/window/rolling.py:2638]: yeah this is materializing things 
prob not a great idea

**Comment by jreback** [comment]
@mroeschke there was no hurry on this one

**Comment by mroeschke** [comment]
Sorry I was too eager on the LGTM. https://github.com/pandas-dev/pandas/pull/46078 reverts this PR.

Yeah this validation is O(number of groups). Happy to have a suggestion regarding a more performant solution.

File: pandas/core/window/rolling.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/window/rolling.py:2638]: self._on is already materialized in __init__. Is that what is meant by materialized?

**Comment by rhshadrach** [comment]
I took another look, I don't see any immediate approaches that would be better to fix the regression. +1 on applying this patch. 

For better performance, can we leverage Index? Adapting the example from https://github.com/pandas-dev/pandas/pull/46065#issuecomment-1046100577

```
def make_df(size):
    dates = pd.date_range(start="2016-01-01 09:30:00", periods=size, freq="s")
    df = pd.DataFrame(
        {
            "A": range(2 * size),
            "B": np.concatenate((dates, dates)),
            "C": np.arange(2 * size),
        }
    )
    return df

df = make_df(5000)
%timeit df.groupby("A").rolling("4s", on="B")
df = make_df(50000)
%timeit df.groupby("A").rolling("4s", on="B")
```

I get

```
179 ms ± 1.73 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
1.82 s ± 48.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Profiling this shows almost all the time spent checking the monotonicity. However
```
df = make_df(50000)
%timeit df.set_index(['A', 'B']).index.is_monotonic_increasing
```

Results in True and I get `11.4 ms ± 114 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`

**Comment by mroeschke** [comment]
> %timeit df.set_index(['A', 'B']).index.is_monotonic_increasing

Interestingly, I guess for a `MultiIndex` only one of the levels needs to be monotonically increasing for it to return True

```
In [3]: df.set_index(['A', 'B']).index
Out[3]:
MultiIndex([(   0, '2016-01-01 09:30:00'),
            (   1, '2016-01-01 09:30:01'),
            (   2, '2016-01-01 09:30:02'),
            (   3, '2016-01-01 09:30:03'),
            (   4, '2016-01-01 09:30:04'),
            (   5, '2016-01-01 09:30:05'),
            (   6, '2016-01-01 09:30:06'),
            (   7, '2016-01-01 09:30:07'),
            (   8, '2016-01-01 09:30:08'),
            (   9, '2016-01-01 09:30:09'),
            ...
            (9990, '2016-01-01 10:53:10'),
            (9991, '2016-01-01 10:53:11'),
            (9992, '2016-01-01 10:53:12'),
            (9993, '2016-01-01 10:53:13'),
            (9994, '2016-01-01 10:53:14'),
            (9995, '2016-01-01 10:53:15'),
            (9996, '2016-01-01 10:53:16'),
            (9997, '2016-01-01 10:53:17'),
            (9998, '2016-01-01 10:53:18'),
            (9999, '2016-01-01 10:53:19')],
           names=['A', 'B'], length=10000)

In [4]: df.set_index(['A', 'B']).index.get_level_values("A").is_monotonic_increasing
Out[4]: True

# This is the rolling axis (on="B")
In [5]: df.set_index(['A', 'B']).index.get_level_values("B").is_monotonic_increasing
Out[5]: False
```

Unfortunately we can't use `.is_monotonic_*` outright because it's the monotonicity of the entire axis while we want to know if each group in the axis is monotonic. e.g.

```
In [6]: print(all(group.B.is_monotonic_increasing for key, group in df.groupby("A")))
True
```

**Pull request merged**